## 1. Neural Network Classifier with **Scikit**

In [16]:
pip install jsonlines

In [17]:
pip install joblib

In [18]:
from sklearn.pipeline import Pipeline
from sklearn.neural_network import MLPRegressor, MLPClassifier
from sklearn.feature_extraction.text import TfidfVectorizer

import time
import numpy as np
from functools import wraps

from sklearn.externals import joblib
from sklearn.model_selection import cross_val_score
import warnings
warnings.filterwarnings('ignore')

import pandas as pd

####  Load JSON Data

In [19]:
import jsonlines
data = []
 
with jsonlines.open('categorized-comments.jsonl') as reader:
    for obj in reader.iter(type=dict, skip_invalid=True):
        data.append(obj)

In [20]:
df = pd.DataFrame(data)

In [21]:
df

cat                                                txt
0            sports  Barely better than Gabbert? He was significant...
1            sports  Fuck the ducks and the Angels! But welcome to ...
2            sports  Should have drafted more WRs.\n\n- Matt Millen...
3            sports            [Done](https://i.imgur.com/2YZ90pm.jpg)
4            sports                                      No!! NOO!!!!!
...             ...                                                ...
912555  video_games  &gt;any chance we can install the entire conte...
912556  video_games             No. It's probably only happened to you
912557  video_games  I think most of the disappointment came from t...
912558  video_games  dishonored 1/2 looked like arse, so what the h...
912559  video_games                                          [removed]

[912560 rows x 2 columns]

In [23]:
df.dropna(inplace = True)

In [22]:
 from sklearn.feature_extraction.text import TfidfVectorizer

In [24]:
TfidfVectorizer = TfidfVectorizer(max_features=2000, stop_words=('english'))

In [25]:
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [ ]:
Remove Stop Words

In [28]:
def tokens(words):
    words = re.sub("[^a-zA-Z]"," ", words)
    text = words.lower().split()                   
    return " ".join(text)

In [ ]:
def numbers(text):
    new_text = []
    for word in text.split():
        if not re.search('\d', word):
            new_text.append(word)
    return ' '.join(new_text)

In [ ]:
df['txt'] = df['txt'].apply(tokens)
df['txt'] = df['txt'].astype(str)

In [ ]:
df['txt'].apply(lambda x: [item for item in x if item not in stop])

In [31]:
df

cat                                                txt
0            sports  barely better than gabbert he was significantl...
1            sports  fuck the ducks and the angels but welcome to a...
2            sports  should have drafted more wrs matt millen probably
3            sports                   done https i imgur com yz pm jpg
4            sports                                             no noo
...             ...                                                ...
912555  video_games  gt any chance we can install the entire conten...
912556  video_games              no it s probably only happened to you
912557  video_games  i think most of the disappointment came from t...
912558  video_games  dishonored looked like arse so what the hell w...
912559  video_games                                            removed

[912557 rows x 2 columns]

####  Split traning and Testing Data

In [32]:
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer

X = df['txt']
y = df['cat']

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.25, random_state = 0)


In [33]:
print(y_train.value_counts())

video_games               542821
sports                    122891
science_and_technology     18704
video_gamesdeo_games           1
Name: cat, dtype: int64


In [34]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import accuracy_score, classification_report
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score

import warnings
warnings.filterwarnings('ignore')

In [35]:
classifer=MLPClassifier(hidden_layer_sizes=[15,10], max_iter=500, random_state=0)

In [38]:
pip install preprocessing

In [42]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import accuracy_score, classification_report
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score
import preprocessing

tfidf = TfidfVectorizer(strip_accents=None, lowercase=False, 
                        max_features=10, 
                        ngram_range=(1,1))
neural_net_pipeline = Pipeline([('vectorizer', tfidf), 
                                ('nn', MLPClassifier(hidden_layer_sizes=(15, 10)))])

neural_net_pipeline.fit(X_train, y_train)

Pipeline(memory=None,
         steps=[('vectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=False, max_df=1.0, max_features=10,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_patte...
                               batch_size='auto', beta_1=0.9, beta_2=0.999,
                               early_stopping=False, epsilon=1e-08,
                               hidden_layer_sizes=(15, 10),
                               learning_rate='constant',
                               learning_r

In [46]:
y_pred = neural_net_pipeline.predict( X_val)

print('Accuracy: {} %'.format(100 * accuracy_score(y_val, y_pred)))

Accuracy: 79.29692294205313 %


## 2. Neural Network Classifier with Keras

In [47]:
from sklearn.datasets import make_multilabel_classification
from sklearn.model_selection import train_test_split
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.losses import binary_crossentropy
from tensorflow.keras.optimizers import Adam

In [ ]:
# Configuration options
n_samples = 10000
n_features = 6
n_classes = 3
n_labels = 2
n_epochs = 50
random_state = 42
batch_size = 250
verbosity = 1
validation_split = 0.2

In [48]:
from sklearn.model_selection import train_test_split
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer

X = df['txt']
y = df['cat']

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.25, random_state = 0)


In [ ]:
X, y = make_multilabel_classification(n_samples=n_samples, n_features=n_features, n_classes=n_classes, n_labels=n_labels, random_state=random_state)
# Split into training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=random_state)


In [ ]:
# Create the model
model = Sequential()
model.add(Dense(32, activation='relu', input_dim=n_features))
model.add(Dense(16, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(n_classes, activation='sigmoid'))

In [ ]:
# Compile the model
model.compile(loss=binary_crossentropy,
              optimizer=Adam(),
              metrics=['accuracy'])

In [ ]:
# Fit data to model
model.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=n_epochs,
          verbose=verbosity,
          validation_split=validation_split)

In [52]:
# Generate generalization metrics
score = model.evaluate(X_test, y_test, verbose=0)
print(f'Test loss: {score[0]} / Test accuracy: {score[1]}')

Epoch 1/50
22/22 [==============================] - 1s 21ms/step - loss: 0.6740 - accuracy: 0.3623 - val_loss: 0.5441 - val_accuracy: 0.4030
Epoch 2/50
22/22 [==============================] - 0s 8ms/step - loss: 0.5262 - accuracy: 0.4326 - val_loss: 0.4908 - val_accuracy: 0.4925
Epoch 3/50
22/22 [==============================] - 0s 8ms/step - loss: 0.4752 - accuracy: 0.5748 - val_loss: 0.4527 - val_accuracy: 0.6269
Epoch 4/50
22/22 [==============================] - 0s 8ms/step - loss: 0.4363 - accuracy: 0.6532 - val_loss: 0.4260 - val_accuracy: 0.6284
Epoch 5/50
22/22 [==============================] - 0s 7ms/step - loss: 0.4090 - accuracy: 0.6554 - val_loss: 0.4047 - val_accuracy: 0.6231
Epoch 6/50
22/22 [==============================] - 0s 7ms/step - loss: 0.3841 - accuracy: 0.6416 - val_loss: 0.3843 - val_accuracy: 0.6276
Epoch 7/50
22/22 [==============================] - 0s 9ms/step - loss: 0.3634 - accuracy: 0.6302 - val_loss: 0.3687 - val_accuracy: 0.5978
Epoch 8/50
22/22 [=

### 3. Classifying Images

In [53]:
pip install intel-tensorflow

In [54]:
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers.convolutional import Conv2D, MaxPooling2D
from keras.utils import np_utils
from keras import backend as K

In [55]:
K.set_image_data_format("channels_first")

In [56]:
np.random.seed(0)

In [57]:
channels = 1
height = 28
width = 28

In [58]:
# data load from mnist dataset
(data_train, target_train), (data_test, target_test) = mnist.load_data()

In [59]:
data_train = data_train.reshape(data_train.shape[0], channels, height, width)

In [60]:
data_test = data_test.reshape(data_test.shape[0],  channels, height, width)

In [61]:
data_train.shape

(60000, 1, 28, 28)

In [62]:
data_test.shape

(10000, 1, 28, 28)

In [63]:
features_train = data_train / 255
features_test = data_test / 255

In [64]:
target_train = np_utils.to_categorical(target_train)
target_test = np_utils.to_categorical(target_test)
number_of_classes = target_test.shape[1]

In [65]:
# Model Creation 

network = Sequential()
network.add(Conv2D(filters=64,
                   kernel_size=(5, 5),
                   input_shape=(channels, width, height),
                   activation='relu',
                   data_format='channels_first'))

In [66]:
network.add(MaxPooling2D(pool_size=(2, 2),padding='same'))

In [67]:
network.add(Dropout(0.5))

In [68]:
network.add(Flatten())

In [69]:
network.add(Dense(128, activation="relu"))

In [70]:
network.add(Dropout(0.5))

In [71]:
network.add(Dense(number_of_classes, activation="softmax"))

In [72]:
network.compile(loss="categorical_crossentropy", 
                optimizer="rmsprop", 
                metrics=["accuracy"])

In [73]:
network.fit(features_train,
            target_train, # Target
            epochs=2, 
            verbose=0, 
            batch_size=1000, 
            validation_data=(features_test, target_test)) 

In [74]:
model_Accuracy=network.evaluate(data_test,target_test,batch_size=200)

50/50 [==============================] - 9s 174ms/step - loss: 12.8369 - accuracy: 0.9701


## Conclusion :

As part of this excercise, Studied the behaviour of convolutional neural network by using Keras. Train and Test dataset.

The Accuracy of the  classify MSINT images using a convolutional neural network is 97%


1. Using the multi-label classifier dataset, fitted a neural network classifier using scikit-learn. The Accuracy is 79%

2. Using the multi-label classifier dataset, fitted a neural network classifier using Keras. The Accuracy is 57%

